src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,329 @@
+            }
+            return false;
+        }
+
+        // For multi-segment parsing of a read only sequence
+        private bool ParseHeadersSlow(
+            ref ReadOnlySequence<byte> buffer,
+            ref KeyValueAccumulator accumulator,
+            ref long headersLength,
+            bool isFinalBlock)
+        {
+            var sequenceReader = new SequenceReader<byte>(buffer);
+            //var position = sequenceReader.Position;
+            //var consumedBytes = default(long);
+
+            while (!sequenceReader.End)
+            {
+                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> line, CrlfDelimiter))
+                {
+                    if (isFinalBlock)
+                    {
+                        throw new InvalidDataException("Unexpected end of Stream, the content may have already been read by another component. ");
+                    }
+
+                    // Don't buffer indefinately
+                    if (headersLength + sequenceReader.Consumed > HeadersLengthLimit)
+                    {
+                        throw new InvalidDataException($"Line length limit {HeadersLengthLimit - headersLength} exceeded.");
+                    }
+
+                    //TODO: keep the sequence and advance the pipeReader in case the sequence length is shorter than the headers limit.
+                }
+
+                if (sequenceReader.Consumed + headersLength > HeadersLengthLimit)
+                {
+                    throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");
+                }
+
+                if (line.IsEmpty)
+                {
+                    buffer = buffer.Slice(sequenceReader.Position);
+                    headersLength += sequenceReader.Consumed;
+                    return true;
+                }
+
+                if (line.IsSingleSegment)
+                {
+                    ParseHeadersFast(line.FirstSpan, ref accumulator, isFinalBlock: true, HeadersLengthLimit - sequenceReader.Consumed - headersLength, out var segmentConsumed);
+                    Debug.Assert(segmentConsumed == line.FirstSpan.Length);
+                    continue;
+                }
+
+                var lineReader = new SequenceReader<byte>(line);
+                ReadOnlySequence<byte> value;
+
+                if (!lineReader.TryReadTo(out var key, ColonDelimiter))
+                {
+                    throw new InvalidDataException($"Invalid header line: {line}");
+                }
+                value = line.Slice(lineReader.Position);
+
+
+                var decodedKey = GetDecodedStringFromReadOnlySequence(key);
+                var decodedValue = GetDecodedStringFromReadOnlySequence(value);
+
+                AppendAndVerify(ref accumulator, decodedKey, decodedValue);
+            }
+
+            buffer = buffer.Slice(sequenceReader.Position);
```

Comment:
Thought that it might interfere with the internal logic of pipeReader.AdvanceTo - I guess that PipeReader handles a default SequencePosition based on what has been received in the readResult.